lib/peer/router.ex

```diff
@@ -0,0 +1,167 @@
+    last_block = Ledger.last_block()
+
+    blocks =
+      if last_block != :err && start <= :binary.decode_unsigned(last_block.index) do
+        start
+        |> Range.new(:binary.decode_unsigned(last_block.index))
+        |> Enum.map(&Ledger.block_at_height/1)
+        |> Enum.filter(& &1 != :none)
+      else
+        []
+      end
+
+    Pico.message(conn, "BLOCK_BATCH_QUERY_RESPONSE", %{blocks: blocks})
+  end
+
+  message "BLOCK_BATCH_QUERY_RESPONSE", %{blocks: blocks} do
+    blocks_count = length(blocks)
+
+    if blocks_count > 0 do
+      Logger.info("Recieved #{blocks_count} blocks from peer.")
+
+      blocks
+      |> Enum.with_index()
+      |> Enum.each(fn {block, i} ->
+        block = Block.sanitize(block)
+
+        if LedgerManager.handle_new_block(block) == :ok do
+          IO.write("Syncing blocks #{round(((i + 1) / blocks_count) * 100)}% [#{i + 1}/#{blocks_count}]\r")
+        end
+      end)
+
+      IO.write("Block Sync Complete")
+    end
+  end
+
+  message "TRANSACTION", transaction do
+    transaction = Transaction.sanitize(transaction)
+    known_transactions = SharedState.get(:transactions) || []
+
+    if Validator.valid_transaction?(transaction) do
+      if transaction not in known_transactions do
+        <<shortid::bytes-size(20), _rest::binary>> = transaction.id
+        Logger.info("Received transaction \e[32m#{shortid}...\e[0m")
+        Pico.broadcast("TRANSACTION", transaction)
+
+        SharedState.set(:transactions, [transaction | known_transactions])
+      end
+    else
+      Logger.info("Received Invalid Transaction. Ignoring.")
+    end
+  end
+
+  message "PORT_RECONNECTION_QUERY", _ do
+    port = Application.get_env(:elixium_core, :port)
+
+    Pico.message(conn, "PORT_RECONNECTION_RESPONSE", %{port: port})
+  end
+
+  message "PORT_RECONNECTION_RESPONSE", %{port: port} do
+    ip = String.to_charlist(handler_state.peername)
+
+    Oracle.inquire(:"Elixir.Elixium.Store.PeerOracle", {:save_known_peer, [{ip, port}]})
+  end
+
+  message "PEER_QUERY_REQUEST", _ do
+    peers =
+      :"Elixir.Elixium.Store.PeerOracle"
+      |> GenServer.call({:load_known_peers, []})
+      |> Enum.take(8)
+
+    Pico.message(conn, "PEER_QUERY_RESPONSE", %{peers: peers})
+  end
+
+  message "PEER_QUERY_RESPONSE", %{peers: peers} do
+    Enum.each(peers, fn peer ->
+      GenServer.call(:"Elixir.Elixium.Store.PeerOracle", {:save_known_peer, [peer]})
+    end)
+  end
+
+  message "NEW_INBOUND_CONNECTION", _ do
+    Pico.message(conn, "PORT_RECONNECTION_QUERY")
+    Pico.message(conn, "PEER_QUERY_REQUEST")
+  end
+
+  message "NEW_OUTBOUND_CONNECTION", _ do
+    # Let's ask our peer for new blocks, if there
+    # are any. We'll ask for all blocks starting from our current index minus
+    # 120 (4 hours worth of blocks before we disconnected) just in case there
+    # was a fork after we disconnected.
+
+    starting_at =
+      case Ledger.last_block() do
+        :err -> 0
```

Comment:
same as above, edge case if needed, or if it's only two cases like before it may be cleaner to use a  ```with-else```

Reply:
With is usually used when chaining results together -- we're not doing that here so I think a case statement is the cleanest way to handle this